### ISSUE TITLE
Support for superseding jobs

### ISSUE THREAD

**Issue by manast** [OptimalBits/bull#848]
It should be possible to add a new job that supersedes a previous job, something like this:

1.- Job 1 is added
2.- Job 1 is active
3.- A job 2 is added
4.- Job 2 should cancel job 1 when added to the queue
5.- Job 2 is active
6.- Job 2 is completed
7.- Queue is empty

**Comment by rafaelamorim** [comment]
Hey @manast, how's going?

That would be awesome indeed!

For me something like:

1- Job with id 'XYZ' is added
2- Job with id 'XYZ' is active
3- Another Job gets created with the same id ('XYZ')
4- Current active Job with id 'XYZ' gets canceled
5- Very new job with id 'XYZ' is active
6- Very new job with id 'XYZ' is completed
7- Queue is empty

Does something like that fits in this same scope that you suggested?

Best! 😃

**Comment by houmark** [comment]
Hey. Any update on this feature? Would really like to see this possible.

**Comment by moltar** [comment]
Isn't that something you can already do with the existing API? It's a bit of work, but the methods are there to introspect the queue and mutate it as needed.

**Comment by houmark** [comment]
@moltar It may be our lack of understanding the API, but we have been unable to do this with the current API. The use case explained by @rafaelamorim explains our need as well.

**Comment by moltar** [comment]
@houmark So _active_ in this case meaning it is currently executing? Then maybe, yeah, it is not possible.

**Comment by houmark** [comment]
@moltar Yes, exactly, problem is we cannot stop/cancel/kill an active job once it's running, so the only real option is to let both run and that creates some unwanted side effects (for example, sometimes the one that started first, finished last but the second one has a newer pay load with data).

**Comment by moltar** [comment]
@houmark I'm new to bull, and was curious if this was possible, which also forced me to learn some bits I did not use yet.

As I understand now, the issues stems from inability to communicate with a running job (e.g. via bull events). There is no way to emit an event to a job in progress that can be listened on inside the process method. I think that would be a useful, and more abstract solution.

I've come up with this proof-of-concept, that uses progress indicator to communicate with the job.

https://gist.github.com/moltar/f9e5843d2c04e3f6e5fd789c18ad3b9c

Technically, this doesn't need to use progress, if you don't care if two jobs run simultaneously for a short time. You could use the `active` event on the queue, and listen to when job 2 gets added (distinguish by name), then search for job 1 and send it an event via emitter (like I am doing with progress right now) to die.

But the biggest problem I am seeing with job abortion, is the cleanup. I imagine scenarios where, for example, I am in the progress of scraping a paginated API and have accumulated a significant amount of data in memory, before committing it to disk/database. This accumulation process uses async requests and are, for example, recursive. How do you ensure this loop exits?

**Comment by manast** [comment]
@moltar thanks for your comment. Basically my idea is that the progress function needs a "cancel/abort" mechanism, similar to some Promise implementations. The main difference would be that the cancellation could also by asynchronous, so a job that is added and is going to superseed another one needs to wait for the current job until it has completed its cancellation.

**Comment by moltar** [comment]
Ok, I tried modifying my solution to handle the waiting part:

https://github.com/moltar/bull-cancel

It is probably incomplete and fragile though, so might need more work.

**Comment by moltar** [comment]
But, as you can see from the log output, even after cancelling the job, the `longOperation` function continues to run until completion. Which is still a major issue, and I don't know what a good solution for that would be?

I tried using promise cancellation feature provided by bluebird, but that doesn't work. When a promise is cancelled, it never resolves nor rejects. The queue promise never gets completed and the job then just hangs until timeout.

I am not aware of any reliable way of cancelling actual operations (rather than the job itself). But I am bit new to Node.js, so if there is a way to mess with the stack and remove the scheduled operations, then I'd like to know how. Promise cancellation was an interesting solution, because, [according to docs](http://bluebirdjs.com/docs/api/cancellation.html), the cancellation goes up the chain, and potentially can cancel the TCP connection or anything else upstream using promise cancellation.

**Comment by manast** [comment]
Yes, I do not think we should use the cancel functionality of bluebird, but gain inspiration from it. For instance, it is convenient to cancel asynchronously. For instance we could have a special method that rejects a promise in the case of cancellation:
```js
queue.process( (job, data) => job.waitForCancel( onCancel => {
  onCancel( () => asyncCleanUps() );
  return mainProcessStuff();
} 
));
```

I am aware cancellation of async tasks is not trivial at all in JS, and this solution has several edge cases but may be good for having a discussion.

**Comment by houmark** [comment]
@manast Checking in here to see if there would be any chance to see this feature implemented in the near future?

Would sponsoring in some way help?

**Comment by manast** [comment]
@houmark sponsoring is always a good incitament :). Currently this feature does not seem to be required by so many users, therefore it has not been prioritised.

**Comment by houmark** [comment]
@manast So how do we get to prioritize it? ;)

**Comment by manast** [comment]
contact me at manuel@optimalbits.com

**Comment by fega** [comment]
@moltar @manast How could be possible with the current API, supersed jobs that are not active?thanks
